### DISCUSSION TITLE
Handling of Udp/Tcp payloads with multiple packages

### DISCUSSION THREAD

**Discussion by jgaulon** [dotpcap/packetnet#153] [Ideas]
Hi,

I'm currently working on a RTCP decoder and I'm facing the following issue.
So far, the design of the class Packet assumes that the payload packet is a singleton, resulting into a chain of packets.
Example: 
RawData -> RawIPPacket -> IPv6Packet -> UDPPacket -> RtcpPacket -> PayloadPacketOrData
or
RawData -> RawIPPacket -> IPv6Packet -> UDPPacket -> DnsPacket

But in some cases / protocols, the payload of UdpPacket (can be TcpPacket too) is actually an array of packets (e.g. 2 RtcpPacket) and not a single one. Additionally, each RtcpPacket has its own payload.

For now, my first Draft makes the second RtcpPacket the PayloadPacketOrData of the first one (which is then its ParentPacket), but this is not really correct neither satisfying. Specially, because the method p.Extract<RtcpPacket>(); returns only a single packet (here the first one) and not an array.

Any idea how do you want to address that?

I see 2 options (below), maybe you have a better idea:

1. option: create a Packet "RtcpPacketContainer" "RtcpPacketList" which actually has no header but only a list/array of "RtcpPacketItem"
2. option: redesign of class "Packet" to turn PayloadPacketOrData into a IList<PayloadPacketOrData>

With the first option, we don't touch the existing code too much, but we can't really do this p.Extract<RtcpPacketItem>();. And we introduce a "Packet" in the chain which is actually not really a Packet.

With the second option, we have to deal with a major refactoring, impacting a lot the existing code and interfaces.

Thanks for feedback and comments.
Julien.

**Comment by jgaulon** [comment]
@PhyxionNL @chmorgan Any comments, ideas, opinion here? Thanks.

**Reply by PhyxionNL** [reply]
To prevent too much breakage, I would prefer 1.
